ext/node/ops/zlib/brotli.rs

```diff
@@ -0,0 +1,237 @@
+
+impl Drop for BrotliCompressCtx {
+  fn drop(&mut self) {
+    unsafe { BrotliEncoderDestroyInstance(self.inst) };
+  }
+}
+
+const ALIGN: usize = std::mem::align_of::<usize>();
+
+fn align_up(size: usize, align: usize) -> usize {
+  (size + align - 1) & !(align - 1)
+}
+
+extern "C" fn brotli_alloc(
+  _: *mut brotli::ffi::broccoli::c_void,
+  size: usize,
+) -> *mut brotli::ffi::broccoli::c_void {
+  fn cvt(ptr: *mut c_void) -> *mut brotli::ffi::broccoli::c_void {
+    // SAFETY: brotli's c_void is repr(u8)
+    unsafe { std::mem::transmute(ptr) }
+  }
+
+  let size =
+    match align_up(size, ALIGN).checked_add(std::mem::size_of::<usize>()) {
+      Some(size) => size,
+      None => return cvt(std::ptr::null_mut()),
+    };
+
+  let layout = match std::alloc::Layout::from_size_align(size, ALIGN) {
+    Ok(layout) => layout,
+    Err(_) => return cvt(std::ptr::null_mut()),
+  };
+
+  // SAFETY: `layout` has non-zero size, guaranteed to be a sentinel address
+  // or a null pointer.
+  unsafe {
+    // Allocate the data, and if successful store the size we allocated
+    // at the beginning and then return an offset pointer.
+    let ptr = alloc::alloc(layout) as *mut usize;
+    if ptr.is_null() {
+      return cvt(ptr as *mut c_void);
+    }
+    *ptr = size;
+    cvt(ptr.add(1) as *mut c_void)
+  }
+}
+
+extern "C" fn brotli_free(
+  _: *mut brotli::ffi::broccoli::c_void,
+  ptr: *mut brotli::ffi::broccoli::c_void,
+) {
+  // SAFETY: `ptr` is a valid pointer to a size, which we can read and then
+  // deallocate.
+  unsafe {
+    if ptr.is_null() {
+      return;
+    }
+    let ptr: *mut usize = std::mem::transmute(ptr);
+    let size = *ptr;
+    let layout = std::alloc::Layout::from_size_align_unchecked(size, ALIGN);
+    alloc::dealloc(ptr as *mut u8, layout);
+  }
+}
+
+#[op]
+pub fn op_create_brotli_compress(
+  state: &mut OpState,
+  params: Vec<(u8, i32)>,
+) -> u32 {
+  let inst = unsafe {
+    BrotliEncoderCreateInstance(
+      Some(brotli_alloc),
+      Some(brotli_free),
+      std::ptr::null_mut(),
+    )
+  };
+
+  for (key, value) in params {
+    unsafe {
+      BrotliEncoderSetParameter(inst, encoder_param(key), value as u32);
+    }
+  }
+
+  state.resource_table.add(BrotliCompressCtx { inst })
+}
+
+fn encoder_param(param: u8) -> BrotliEncoderParameter {
+  // SAFETY: BrotliEncoderParam is valid for 0-255
+  unsafe { std::mem::transmute(param as u32) }
+}
+
+#[op]
+pub fn op_brotli_compress_stream(
+  state: &mut OpState,
+  rid: u32,
+  input: &[u8],
+  output: &mut [u8],
+) -> Result<usize, AnyError> {
+  let ctx = state.resource_table.get::<BrotliCompressCtx>(rid)?;
+
+  unsafe {
+    let mut available_in = input.len();
+    let mut next_in = input.as_ptr();
+    let mut available_out = output.len();
+    let mut next_out = output.as_mut_ptr();
+    let mut total_out = 0;
+
+    if BrotliEncoderCompressStream(
+      ctx.inst,
+      BrotliEncoderOperation::BROTLI_OPERATION_PROCESS,
```

Comment:
This operation may not flush to the output buffer, so you will probably need to call the FLUSH op too to actually move the data into the output buffers. See my other Brotli PR.